### PR TITLE
Add a new optimization to version loops containing loads using block ptrs with unknown strides

### PR DESCRIPTION
**Pull request by etiotto**
This PR version `scf.for` loops containing tt.load on block pointers, where the block ptr is declared by a make_tensor_ptr operation with no strides equal to 1 (strides unknown at compile time).  The versioned loop will then contain tt.load operations that can ber marked as "row-major/column-major" and eventually lowered to efficient 2D block IO loads.

### CODE REVIEW

**Comment by chengjunlu** [comment]
LGTM. 
The better we can make the version condition to be a SYCL constant to reduce complication in the kernel binary in the later changes.
This also can be used for some other case like pitch < base width.

**Comment by whitneywhtsang** [comment]
Tensor descriptor has a restriction that the last dimension stride must be 1: https://github.com/intel/intel-xpu-backend-for-triton/blob/main/python/triton/language/semantic.py#L1946.
We have relaxed this restriction on Intel backend for tutorial 6 fp8 handling. Wonder if we should change it back to be consistent with other backends.

File: third_party/intel/include/Dialect/Triton/Transforms/Passes.td [hunk 1]
```diff
@@ -107,4 +107,39 @@ def TritonIntelRemoveBoundaryChecks
   ];
 }
 
+def TritonIntelStrideVersioning
+    : Pass<"triton-intel-stride-versioning", "mlir::ModuleOp"> {
+  let summary = "Version loops containing block pointers loads if none of the is 1";
```
Comment (whitneywhtsang) [inline third_party/intel/include/Dialect/Triton/Transforms/Passes.td]:
```suggestion
  let summary = "Version loops containing block pointers loads if none of them is 1";
```

File: third_party/intel/include/Dialect/Triton/Transforms/Passes.td [hunk 2]
```diff
@@ -107,4 +107,39 @@ def TritonIntelRemoveBoundaryChecks
...
+      %load = tt.load %ptr : !tt.ptr<tensor<512x64xf16>>
+      ...
+    }
+
+    The transformation creates:
+
+    %ptr = tt.make_tensor_ptr %base, [%s0, %s1], [%a, %b], [%x, %y] : <tensor<512x64xf16>>
+    if (%b == 1)
+      scf.for ... {
+        %load = tt.load %ptr : !tt.ptr<tensor<512x64xf16>>
```
Comment (whitneywhtsang) [inline third_party/intel/include/Dialect/Triton/Transforms/Passes.td]: is the intention to have the load changed here?

File: third_party/intel/lib/Utils/Utility.cpp [hunk 3]
```diff
@@ -182,6 +182,11 @@ Value getFinalValue(Value value) {
     return divOp.getResult();
   }
 
+  if (auto extOp = dyn_cast<arith::ExtSIOp>(defOp))
+    return getFinalValue(extOp.getIn());
+  if (auto extOp = dyn_cast<arith::ExtUIOp>(defOp))
+    return getFinalValue(extOp.getIn());
```
Comment (whitneywhtsang) [inline third_party/intel/lib/Utils/Utility.cpp:188]: is this supported?
```suggestion
  if (auto extOp = dyn_cast<arith::ExtSIOp, arith::ExtUIOp>(defOp))
    return getFinalValue(extOp.getIn());
```

File: third_party/intel/lib/Dialect/Triton/Transforms/StrideVersioning.cpp [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+          Value ptr = loadOp.getPtr();
+          if (!tt::isTensorPointerType(ptr.getType()))
+            return false;
+
+          auto tensorType = cast<RankedTensorType>(
+              cast<tt::PointerType>(ptr.getType()).getPointeeType());
+          if (tensorType.getRank() > 2)
+            return false;
+
+          auto isOneConst = [](Value v) {
```
Comment (whitneywhtsang) [inline third_party/intel/lib/Dialect/Triton/Transforms/StrideVersioning.cpp]: IMO isOne or isConstOne is a better name, as isOneConst may mean the Value is a constant, and not a vector of constants.
```suggestion
          auto isOne = [](Value v) {
```

File: third_party/intel/lib/Dialect/Triton/Transforms/StrideVersioning.cpp [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+            return false;
+          };
+
+          // If no stride has value equal to one we have found a candidate
+          // operation.
+          auto makeTensorPtrOp = tt::getMakeTensorPtrOp(ptr);
+          OperandRange strides = makeTensorPtrOp.getStrides();
+
+          bool isCandidate = true;
+          for (Value stride : strides) {
```
Comment (whitneywhtsang) [inline third_party/intel/lib/Dialect/Triton/Transforms/StrideVersioning.cpp]: do you think any_of would be a good use here?

File: third_party/intel/lib/Dialect/Triton/Transforms/StrideVersioning.cpp [hunk 6]
```diff
@@ -0,0 +1,273 @@
...
+                  assert(tt::isTensorPointerType(ptr.getType()) &&
+                         "Expecting a block ptr");
+
+                  auto makeTensorPtrOp = tt::getMakeTensorPtrOp(ptr);
+                  makeTensorPtrOps.push_back(makeTensorPtrOp);
+                  OperandRange strides = makeTensorPtrOp.getStrides();
+
+                  for (Value stride : strides) {
+                    Value finalVal = tt::intel::getFinalValue(stride);
+                    assert(finalVal && "Expecting a valid mask");
```
Comment (whitneywhtsang) [inline third_party/intel/lib/Dialect/Triton/Transforms/StrideVersioning.cpp]: valid stride?

File: third_party/intel/lib/Dialect/Triton/Transforms/StrideVersioning.cpp [hunk 7]
```diff
@@ -0,0 +1,273 @@
...
+                    Operation *defOp = finalVal.getDefiningOp();
+                    if (defOp)
+                      continue;
+
+                    auto blockArg = cast<BlockArgument>(finalVal);
+                    Operation *parentOp = blockArg.getOwner()->getParentOp();
+                    // TODO: handle block args for things that aren't a
+                    // function.
+                    auto funcOp = cast<tt::FuncOp>(parentOp);
+                    auto divisibilityAttr =
```
Comment (whitneywhtsang) [inline third_party/intel/lib/Dialect/Triton/Transforms/StrideVersioning.cpp:245]: why divisibility matters here?

File: third_party/intel/lib/Dialect/Triton/Transforms/StrideVersioning.cpp [hunk 8]
```diff
@@ -0,0 +1,273 @@
...
+                        funcOp.getArgAttrOfType<IntegerAttr>(
+                            blockArg.getArgNumber(), "tt.divisibility");
+                    if (divisibilityAttr) {
+                      assert(divisibilityAttr.getValue().isStrictlyPositive() &&
+                             "Unexpected divisibility value");
+                      continue;
+                    }
+
+                    makeTensorPtrToStride[makeTensorPtrOp] = blockArg;
```
Comment (whitneywhtsang) [inline third_party/intel/lib/Dialect/Triton/Transforms/StrideVersioning.cpp:255]: `makeTensorPtrToStride[makeTensorPtrOp]` is going to be override to the last stride that matches the condition? Should we try with row major first?

File: third_party/intel/lib/Dialect/Triton/Transforms/StrideVersioning.cpp (hunk 7 above)
Comment (etiotto) [reply third_party/intel/lib/Dialect/Triton/Transforms/StrideVersioning.cpp:245]: The reason it matters is that if a stride is an argument that has the divisibility attribute of say 16, than that argument may only assume values that ate multiple of 16 and therefore it cannot be 1.

File: third_party/intel/lib/Dialect/Triton/Transforms/StrideVersioning.cpp (hunk 8 above)
Comment (etiotto) [reply third_party/intel/lib/Dialect/Triton/Transforms/StrideVersioning.cpp:255]: ok we can traverse the strides in a specific order, from the innermost to the outermost dimension as specified by the order attribute.

File: third_party/intel/lib/Utils/Utility.cpp (hunk 3 above)
Comment (etiotto) [reply third_party/intel/lib/Utils/Utility.cpp:188]: No it's not supported.

File: third_party/intel/include/Dialect/Triton/Transforms/Passes.td (hunk 1 above)
Comment (etiotto) [reply third_party/intel/include/Dialect/Triton/Transforms/Passes.td]: done

File: third_party/intel/include/Dialect/Triton/Transforms/Passes.td (hunk 2 above)
Comment (etiotto) [reply third_party/intel/include/Dialect/Triton/Transforms/Passes.td]: The load has changed, it uses a block ptr that is generated by a new make_tensor_ptr operation where the stride is one.

File: third_party/intel/lib/Dialect/Triton/Transforms/StrideVersioning.cpp (hunk 4 above)
Comment (etiotto) [reply third_party/intel/lib/Dialect/Triton/Transforms/StrideVersioning.cpp]: done

File: third_party/intel/lib/Dialect/Triton/Transforms/StrideVersioning.cpp (hunk 6 above)
Comment (etiotto) [reply third_party/intel/lib/Dialect/Triton/Transforms/StrideVersioning.cpp]: fixed typo

File: third_party/intel/lib/Dialect/Triton/Transforms/StrideVersioning.cpp (hunk 7 above)
Comment (etiotto) [reply third_party/intel/lib/Dialect/Triton/Transforms/StrideVersioning.cpp:245]: added a comment to the code.

File: third_party/intel/lib/Dialect/Triton/Transforms/StrideVersioning.cpp (hunk 5 above)
Comment (etiotto) [reply third_party/intel/lib/Dialect/Triton/Transforms/StrideVersioning.cpp]: changed to use none_of

File: third_party/intel/lib/Dialect/Triton/Transforms/StrideVersioning.cpp (hunk 8 above)
Comment (whitneywhtsang) [reply third_party/intel/lib/Dialect/Triton/Transforms/StrideVersioning.cpp:255]: should we stop the iteration when `makeTensorPtrToStride[makeTensorPtrOp]` is defined once?

Comment (etiotto) [reply third_party/intel/lib/Dialect/Triton/Transforms/StrideVersioning.cpp:255]: fixed

File: third_party/intel/lib/Dialect/Triton/Transforms/StrideVersioning.cpp [hunk 9]
```diff
@@ -0,0 +1,272 @@
...
+#define GEN_PASS_DEF_TRITONINTELSTRIDEVERSIONING
+#include "intel/include/Dialect/Triton/Transforms/Passes.h.inc"
+} // namespace mlir::triton::intel
+
+namespace {
+
+class OpSelector {
+public:
+  // Select load operations that use a block ptr defined by a
+  // make_tensor_ptr operation with at least one stride that has unknown value
```
Comment (whitneywhtsang) [inline third_party/intel/lib/Dialect/Triton/Transforms/StrideVersioning.cpp:35]: I think the comment is good, can we ensure at least one stride has unknown compile time value in the call?

**Review by whitneywhtsang — APPROVED** [review]

File: third_party/intel/lib/Dialect/Triton/Transforms/StrideVersioning.cpp (hunk 9 above)
Comment (etiotto) [reply third_party/intel/lib/Dialect/Triton/Transforms/StrideVersioning.cpp:35]: ok let me add a test case to ensure that if all strides have known values (constants) and none of them is one we do not version.
